### PR TITLE
[SPARK-42538][CONNECT] Make `sql.functions#lit` function support more types

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
This pr aims add more types support of `sql.functions#lit` function, include:

- Decimal
- Instant
- Timestamp
- LocalDateTime
- Date
- Duration
- Period
- CalendarInterval

### Why are the changes needed?
Make ·sql.functions#lit· function support more types

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?

- Add new test
- Manual checked new case with Scala-2.13

### CODE REVIEW

**Comment by LuciferYang** [comment]
Some other things to do, will continue tomorrow

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala [hunk 1]
```diff
@@ -136,10 +148,51 @@ object functions {
...
+      case v: Duration => createLiteral(_.setDayTimeInterval(IntervalUtils.durationToMicros(v)))
+      case v: Period => createLiteral(_.setYearMonthInterval(IntervalUtils.periodToMonths(v)))
+      case v: CalendarInterval => createCalendarIntervalLiteral(v.months, v.days, v.microseconds)
       case null => createLiteral(_.setNull(nullType))
+      case v: Literal => literalToColumn(v)
       case _ => unsupported(s"literal $literal not supported (yet).")
     }
   }
+
+  private def literalToColumn(literal: Literal): Column = {
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala:167]: We don't need this. In fact for the next Spark release we will be removing the Catalyst dependency.

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala [hunk 2]
```diff
@@ -136,10 +148,51 @@ object functions {
       case v: Array[Byte] => createLiteral(_.setBinary(ByteString.copyFrom(v)))
       case v: collection.mutable.WrappedArray[_] => lit(v.array)
       case v: LocalDate => createLiteral(_.setDate(v.toEpochDay.toInt))
+      case v: UTF8String => createLiteral(_.setString(v.toString))
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala:151]: This is internal API. Can you remove it.

**Comment by hvanhovell** [comment]
@LuciferYang thanks for the PR! Which datatypes are we still missing? I think we still some collection support?

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala (hunk 1 above)
Comment (LuciferYang) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala:167]: So, we should remove `case v: Literal => literalToColumn(v)` ?

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala (hunk 2 above)
Comment (LuciferYang) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala:151]: OK

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala (hunk 1 above)
Comment (hvanhovell) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala:167]: Yeah we should. Sorry about that.

**Comment by LuciferYang** [comment]
I refer 

https://github.com/apache/spark/blob/87e3d5625e76bb734b8dd753bfb25002822c8585/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala#L64-L102

the missing is `case a: Array[_]`

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala (hunk 1 above)
Comment (LuciferYang) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala:167]: ok, let me update the code

**Comment by hvanhovell** [comment]
For Array I guess we can just make a nested call to lit for each element. No need to get CatalystTypeConverters involved. You get bonWe do have to check what `CatalystTypeConverters` currently supports though.

**Comment by hvanhovell** [comment]
Oh and if it becomes too large I am fine with merging this first, and doing array in a follow-up.

**Comment by LuciferYang** [comment]
hmm... If I understand correctly, the current Literal does not support any collection type? Do we need to add some message types to support them?

https://github.com/apache/spark/blob/87e3d5625e76bb734b8dd753bfb25002822c8585/connector/connect/common/src/main/protobuf/spark/connect/expressions.proto#L149-L175

**Comment by hvanhovell** [comment]
lol, no it does not. Let's just implement what we support, and do the rest in a different PR.

**Comment by LuciferYang** [comment]
OK

**Comment by LuciferYang** [comment]
> Oh and if it becomes too large I am fine with merging this first, and doing array in a follow-up.

I hope we can join this pr first if no other need to change. In addition, I need to go to bed as soon as possible. It's 1:00 in my time zone :)

**Comment by hvanhovell** [comment]
go to sleep!

**Comment by LuciferYang** [comment]
@hvanhovell Is there anything else can help Scala Client? @panbingkun told me that he also wanted to take some work related to connect.

**Comment by hvanhovell** [comment]
@LuciferYang @panbingkun that would be great! I will create an epic, with a bunch of todo's.

**Comment by LuciferYang** [comment]
Is there anything else need change this pr?

**Comment by hvanhovell** [comment]
@LuciferYang can you update your PR?

**Review by hvanhovell — APPROVED** [review]
LGTM

**Comment by hvanhovell** [comment]
@LuciferYang @panbingkun I created an epic with a bunch of things you can pick up: https://issues.apache.org/jira/browse/SPARK-42554

**Comment by amaliujia** [comment]
LGTM but please rebase this PR to solve conflict.

**Comment by hvanhovell** [comment]
Merging to master/3.4. Thanks!

**Pull request closed without merging**